File: framework/languages/java/e2e/InstanceSpot/Owner/src/main/java/systems/zlink/e2e/instancespot/owner/ProbeSpot.java
package systems.zlink.e2e.instancespot.owner;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import systems.zlink.framework.spots.ZLinkInstanceSpot;
import systems.zlink.framework.spots.ZLinkInstanceSpotContext;
import systems.zlink.framework.spots.ZLinkSpotClosingContext;

public final class ProbeSpot implements ZLinkInstanceSpot {
    private final ZLinkInstanceSpotContext context;
    private final EvidenceStore evidence;
    private final GateController gates;
    private final AtomicLong handlerSequence = new AtomicLong();
    private final AtomicInteger activeHandlers = new AtomicInteger();

    public ProbeSpot(
        ZLinkInstanceSpotContext context,
        EvidenceStore evidence,
        GateController gates) {
        this.context = context;
        this.evidence = evidence;
        this.gates = gates;
        evidence.record(
            "FACTORY",
            context.spotId(),
            "",
            "",
            context.objectGeneration(),
            0,
            "constructor");
    }

    @Override
    public ZLinkInstanceSpotContext context() {
        return context;
    }

    EvidenceStore evidence() {
        return evidence;
    }

    GateController gates() {
        return gates;
    }

    long nextHandlerSequence() {
        return handlerSequence.incrementAndGet();
    }

    int enterHandler(String operationId, String payload) {
        int active = activeHandlers.incrementAndGet();
        evidence.record(
            "HANDLER_ENTER",
            context.spotId(),
            operationId,
            payload,
            context.objectGeneration(),
            active,
            "request");
        return active;
    }

    void leaveHandler(String operationId, String payload) {
        int active = activeHandlers.decrementAndGet();
        evidence.record(
            "HANDLER_COMMIT",
            context.spotId(),
            operationId,
            payload,
            context.objectGeneration(),
            active,
            "domain-commit");
    }

    @Override
    public void configure() {
        context.handlers().addPacket(ProbeSendHandler.class);
        context.handlers().addPacket(CloseHandler.class);
    }

    @Override
    public CompletionStage<Void> onInitialize() {
        evidence.record(
            "INITIALIZE",
            context.spotId(),
            "",
            "",
            context.objectGeneration(),
            0,
            "ready");
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onClosing(ZLinkSpotClosingContext closing) {
        evidence.record(
            "CLOSING",
            context.spotId(),
            "",
            "",
            context.objectGeneration(),
            0,
            closing.reason().name());
        return CompletableFuture.completedFuture(null);
    }
}

File: framework/languages/java/e2e-kotlin/InstanceSpot/Owner/src/main/kotlin/systems/zlink/e2e/kotlin/instancespot/owner/ProbeSpot.kt
package systems.zlink.e2e.kotlin.instancespot.owner

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import systems.zlink.e2e.kotlin.instancespot.shared.Contracts
import systems.zlink.framework.kotlin.ZLinkSuspendingInstanceSpot
import systems.zlink.framework.kotlin.addHandler
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.spots.ZLinkInstanceSpotContext
import systems.zlink.framework.spots.ZLinkSpotPacketHandler
import systems.zlink.framework.spots.ZLinkSpotRequestHandler
import systems.zlink.framework.spots.ZLinkSpotClosingContext

class ProbeSpot(
    override val context: ZLinkInstanceSpotContext,
    private val evidence: EvidenceStore,
    private val gates: GateController,
) : ZLinkSuspendingInstanceSpot() {
    private val handlerSequence = AtomicLong()
    private val activeHandlers = AtomicInteger()

    init {
        evidence.record(
            "FACTORY", context.spotId(), generation = context.objectGeneration(), detail = "kotlin-constructor",
        )
    }

    override fun configure() {
        context.handlers().addHandler<ProbeSendHandler>()
        context.handlers().addHandler<CloseHandler>()
    }

    override suspend fun onInitializeSuspending() {
        evidence.record("INITIALIZE", context.spotId(), generation = context.objectGeneration(), detail = "ready")
    }

    override suspend fun onClosingSuspending(closing: ZLinkSpotClosingContext) {
        evidence.record(
            "CLOSING", context.spotId(), generation = context.objectGeneration(), detail = closing.reason().name,
        )
    }

    fun enter(operationId: String, payload: String): Int = activeHandlers.incrementAndGet().also {
        evidence.record(
            "HANDLER_ENTER", context.spotId(), operationId, payload,
            context.objectGeneration(), it, "request",
        )
    }

    fun commit(operationId: String, payload: String) {
        val active = activeHandlers.decrementAndGet()
        evidence.record(
            "HANDLER_COMMIT", context.spotId(), operationId, payload,
            context.objectGeneration(), active, "domain-commit",
        )
    }

    fun awaitPayload(payload: String) = gates.awaitPayload(payload)
    fun awaitGate(gateId: String) = gates.await(gateId)
    fun nextSequence() = handlerSequence.incrementAndGet()
    fun evidence() = evidence
}

class ProbeRequestHandler : ZLinkSpotRequestHandler<ProbeSpot, Contracts.InstanceRequest, Contracts.InstanceReply> {
    override fun handle(
        spot: ProbeSpot,
        request: Contracts.InstanceRequest,
    ): CompletionStage<Contracts.InstanceReply> {
        spot.enter(request.operationId, request.payload)
        return spot.awaitPayload(request.payload).thenApply {
            Contracts.InstanceReply(
                spot.context.spotId(), request.operationId, request.payload,
                spot.evidence().snapshot().rid, spot.evidence().snapshot().lifecycleId,
                spot.context.objectGeneration(), spot.nextSequence(),
            ).also { spot.commit(request.operationId, request.payload) }
        }
    }
}

class ProbeSendHandler : ZLinkSpotPacketHandler<ProbeSpot, Contracts.InstanceSend> {
    override fun handle(spot: ProbeSpot, message: Contracts.InstanceSend): CompletionStage<Void> =
        spot.awaitPayload(message.payload).thenRun {
            spot.evidence().record(
            "SEND_HANDLER", spot.context.spotId(), message.operationId, message.payload,
            spot.context.objectGeneration(), detail = "one-way",
            )
        }
}

class CloseHandler : ZLinkSpotPacketHandler<ProbeSpot, Contracts.CloseRequest> {
    override fun handle(spot: ProbeSpot, message: Contracts.CloseRequest): CompletionStage<Void> =
        spot.awaitGate(message.gateId).thenCompose {
            spot.context.close().thenAccept { closed ->
                spot.evidence().record(
            "CLOSE_RESULT", spot.context.spotId(), message.operationId,
            generation = spot.context.objectGeneration(), detail = closed.toString(),
                )
            }
        }
}
